digicampipe/image/utils.py

```diff
@@ -0,0 +1,416 @@
+from astropy.io import fits
+from astropy.stats import sigma_clipped_stats
+import numpy as np
+import scipy
+from skimage.draw import polygon
+import matplotlib.pyplot as plt
+from time import clock
+import cv2
+from scipy import signal, ndimage, optimize
+import numbers
+from photutils import DAOStarFinder
+
+
+def crop_image(image, crop_pixel1, crop_pixel2):
+    """
+    :param image: 2D array of the image data with a shape = (vertical size, horizontal size)
+    :param crop_pixel1: iterable of 2 ints: position (pos_x , pos_y) of one corner of the returned region
+    :param crop_pixel2: iterable of 2 ints: position (pos_x , pos_y) of the opposite corner of the returned region
+    :return: the cropped image, the 1st corner and the opposite corner positions
+    """
+    if type(image) is str:
+        image = fits.open(image)[0].data
+    if type(image) is not np.ndarray:
+        raise AttributeError("image must be a filename or a numpy array.")
+    if crop_pixel1 is not list and len(crop_pixel1) is not 2:
+        raise AttributeError("crop_pixel1 must be a list of 2 ints", "(position in px of one corner of the crop area).")
+    if crop_pixel2 is not list and len(crop_pixel1) is not 2:
+        raise AttributeError("crop_pixel2 must be a list of 2 ints",
+                             "(position in px of the other corner of the crop area).")
+    crop_xmin, crop_ymin = crop_pixel1
+    crop_xmax, crop_ymax = crop_pixel2
+    if not isinstance(crop_xmin, numbers.Integral) or not isinstance(crop_ymin, numbers.Integral):
+        print('Warning in crop_image(): crop_pixel1 does not contain ints')
+        crop_xmin = int(crop_xmin)
+        crop_ymin = int(crop_ymin)
+    if not isinstance(crop_xmax, numbers.Integral) or not isinstance(crop_ymax, numbers.Integral):
+        print('Warning in crop_image(): crop_pixel2 does not contain ints')
+        crop_xmax = int(crop_xmax)
+        crop_ymax = int(crop_ymax)
+    npixel_y, npixel_x = image.shape
+    if not (0 <= crop_xmin < npixel_x and 0 <= crop_ymin < npixel_y):
+        raise AttributeError("invalid crop_pixel position:", crop_pixel1, crop_pixel2, "image size is: ", image.shape)
+    if not (0 < crop_xmax < npixel_x and 0 < crop_ymax < npixel_y):
+        raise AttributeError("invalid crop_pixel position:", crop_pixel1, crop_pixel2, "image size is: ", image.shape)
+    if crop_xmin > crop_xmax:
+        crop_xmin, crop_xmax = crop_xmax, crop_xmin
+    if crop_ymin > crop_ymax:
+        crop_ymin, crop_ymax = crop_ymax, crop_ymin
+    crop_pixel1 = crop_xmin, crop_ymin
+    crop_pixel2 = crop_xmax, crop_ymax
+    return image[crop_ymin:crop_ymax, crop_xmin:crop_xmax], crop_pixel1, crop_pixel2
```

Comment:
why does this function also return the rectangle again? I mean sure, it considers boundary effects, but is this actually needed anywhere? 

As a more general thought. Functions can return tuples in python, but often people cannot remember what the order of the stuff inside the tuple is. So often functions returning tuples are hard to use, and if they are used somewhere in code, it is hard to understand what they return, if the variable names are not chosen very good.

Reply:
The rectangle is returned as it can differ from the given one (f.e. if the rectangle is partially out of the image).
It is then use to translate coordinates from the cropped region to the full image.
I do not like much returning tuples either, but creating a class for that one seemed a bit overkill.

Reply:
I did it anyway in 4724d7158293